### PR TITLE
SFRestAPI Upgrades, Ability to invoke endpoints in an unauthenticated context

### PR DESCRIPTION
**Pull request by trooper2013**
@bhariharan @wmathurin please review. This PR Does the following. 

Note: All Core,SmartStore & SmartSync tests are passing. Im still testing all sample apps to ensure that the result is as expected.

1. Fixes issues related to community files api(s). [Read about the required changes](https://developer.salesforce.com/docs/atlas.en-us.chatterapi.meta/chatterapi/connect_resources_files_rendition.htm).

2. Adds a sharedGlobalInstance of SFRestAPI for use in unauthenticated context.

3. Allows ability to invoke custom REST API endpoints.

```
 SFRestRequest *request = [SFRestRequest customUrlRequestWithMethod:SFRestMethodGET baseURL:@"https://api.github.com" path:@"/orgs/forcedotcom/repos" queryParams:nil];
 request.requiresAuthentication = NO;  
 [[SFRestAPI sharedGlobalInstance] sendRESTRequest:request failBlock:^(NSError *  e, 
      NSURLResponse * rawResponse) {
       
  } completeBlock:^(id  resp, NSURLResponse *  rawResponse) {
           NSDictionary *response = resp;
           // response should have forcedotcom repos      
  }];
```
4. Allows the setting of service type in Request. This enables the OS to prioritize requests.  
[Read](https://developer.apple.com/documentation/foundation/nsurlrequestnetworkservicetype?language=objc).

5. Added Unit tests for most of the changes.

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.h [hunk 1]
```diff
@@ -38,6 +38,17 @@ typedef NS_ENUM(NSInteger, SFRestMethod) {
     SFRestMethodPATCH
 } NS_SWIFT_NAME(RestRequest.Method);
 
+/**
+ * HTTP methods for requests.
+ */
+typedef NS_ENUM(NSInteger, SFSDKNetworkServiceType) {
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.h]: Would it make sense for this to live in `SFNetwork` instead since it's more of a network property than API client related?

Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.h]: This is part of NSURLRequest in foundation. Hence it made more sense to add it our SFRestRequest. i.e. easier to lookup and use for a developer.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 2]
```diff
@@ -272,7 +289,7 @@ - (void)enqueueRequest:(SFRestRequest *)request delegate:(id<SFRestDelegate>)del
                 [strongSelf notifyDelegateOfResponse:delegate request:request data:dataForDelegate rawResponse:response];
             }
             // 401 (and sometimes 403) indicates refresh is required.
-            else if (request.shouldRefreshOn403 ? (statusCode == 401 || statusCode == 403) : (statusCode == 401)) {
+            else if (request.requiresAuthentication && request.shouldRefreshOn403 ? (statusCode == 401 || statusCode == 403) : (statusCode == 401)) {
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: Or maybe we should make sure shouldRetry is no if requiresAuthenticate is false.
By changing 
```objective-c
- (void)send:(SFRestRequest *)request delegate:(id<SFRestDelegate>)delegate {
    [self send:request delegate:delegate shouldRetry:YES];
}
```
to
```objective-c
- (void)send:(SFRestRequest *)request delegate:(id<SFRestDelegate>)delegate {
    if (request.requiresAuthentication) {
        [self send:request delegate:delegate shouldRetry:YES];
    }
}
```

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI+Files.m [hunk 3]
```diff
@@ -40,41 +40,41 @@
 @implementation SFRestAPI (Files)
 
 - (SFRestRequest *) requestForOwnedFilesList:(NSString *)userId page:(NSUInteger)page {
-    NSString *path = [NSString stringWithFormat:@"/%@/connect/files/users/%@", self.apiVersion, (userId == nil ? ME : userId)];
+    NSString *path = [NSString stringWithFormat:@"/%@/connect%@/files/users/%@", self.apiVersion,[self communitiesUrlPathIfRequired],  (userId == nil ? ME : userId)];
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI+Files.m]: Ideally the community bug fix should be in a separate PR since it's completely unrelated to the support for unauthenticated calls.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.h (hunk 1 above)
Comment (bhariharan) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.h]: Sounds good.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m [hunk 4]
```diff
@@ -39,7 +39,7 @@ - (id)initWithMethod:(SFRestMethod)method serviceHostType:(SFSDKRestServiceHostT
         self.baseURL = baseURL;
         self.path = path;
         self.queryParams = [queryParams mutableCopy];
-        self.endpoint = kSFDefaultRestEndpoint;
+        self.endpoint = (hostType == SFSDKRestServiceHostTypeCustom)?@"":kSFDefaultRestEndpoint;
```
[the remark this answers is not in the document]
Comment (bhariharan) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m]: `baseUrl` doesn't always have the information of whether an API should go against `loginUrl` or `instanceUrl`. I think we still need `hostType` for those special `loginUrl` URLs like the `userInfo` endpoint. I forget the exact specifics.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m [hunk 5]
```diff
@@ -122,84 +126,83 @@ - (void)setCustomRequestBodyStream:(NSInputStream* (^)(void))bodyInputStreamBloc
...
 
 - (NSURLRequest *)prepareRequestForSend:(SFUserAccount *)user {
-    if (user) {
-
-        /*
-         * If an absolute URL is passed in, use it as-is. If a relative URL is passed in,
-         * parse it and put the pieces together to construct the full URL.
-         */
-        NSMutableString *fullUrl = nil;
-        if ([[self.path lowercaseString] hasPrefix:@"https://"]) {
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m]: @trooper2013 We still need this `if` condition as I mentioned in our conversation offline. There are a few different use cases for this. Some are covered by your new method that takes a custom endpoint (the native use cases), but removing this will break existing clients that rely on this behavior (we have some users internally). Also, it will remove the ability to accomplish this from our `Cordova` and `ReactNative` apps. We need to bring this `if` condition back, at least for now.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m [hunk 6]
```diff
@@ -122,84 +126,83 @@ - (void)setCustomRequestBodyStream:(NSInputStream* (^)(void))bodyInputStreamBloc
...
+
+    //Set the service host type
+    NSURLRequestNetworkServiceType serviceType = [self urlRequestServiceType:self.networkServiceType];
+    [self.request setNetworkServiceType:serviceType];
+    
+    // Sets HTTP method on the request.
+    [self.request setHTTPMethod:[SFRestRequest httpMethodFromSFRestMethod:self.method]];
+
+    // Sets OAuth Bearer token header on the request (if not already present).
+    if (self.requiresAuthentication && ![self.request.allHTTPHeaderFields.allKeys containsObject:@"Authorization"]) {
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m]: This should be:
```
if (user && self.requiresAuthentication && ![self.request.allHTTPHeaderFields.allKeys containsObject:@"Authorization"])
```
This guards against someone who has an unauthenticated instance (`user == nil`), but has `self.requiresAuthentication == YES`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestRequest.m]: Looking at `prepareRequestForSend`, it looks like if path starts with https then it is used the full url. In that case, base url, end point and host type don't matter. 

Maybe the method `customUrlRequestWithMethod` should just take one parameter: the full url (which would be fed into the path field of the SFRestRequest).

Also the comment for the method is misleading since it doesn't help the case where you want to hit a custom end point on salesforce (e.g. an api under `/services/apexrest`). For that case, a useful factory method would be one taking end point and path.

So in summary, useful new factory methods would be:
```objective-c
+ (instancetype)customUrlRequestWithMethod:(SFRestMethod)method fullURL:(NSString *)fullURL queryParams:(nullable NSDictionary<NSString*, id> *)queryParams;

+ (instancetype)customEndPointRequestWithMethod:(SFRestMethod)method endPoint:(NSString *)endPoint path:(NSString *)path queryParams:(nullable NSDictionary<NSString*, id> *)queryParams;
```
